trestle remove and testing, with coverage at 100%

jeffdmgit opened the pull request:
## Types of changes

- \[ \] Bug fix (non-breaking change which fixes an issue)
- \[x\] New feature (non-breaking change which adds functionality)
- \[ \] Breaking change (fix or feature that would cause existing functionality to change)
- \[x\] My code follows the code style of this project.
- \[ \] My change requires a change to the documentation.
- \[ \] I have updated the documentation accordingly.
- \[x\] I have added tests to cover my changes.
- \[x\] All new and existing tests passed.
- \[ \] All commits are signed.

* Adds **trestle remove** functionality.
* Includes 100% coverage for testing it.

This PR closes #66.

butler54 commented on trestle/core/commands/remove.py:
```diff
@@ -15,10 +15,113 @@
+                # The element already exists
+                if type(deleting_element) is list:
+                    pass
+                    warnings.warn('trestle remove does not support removing elements of a list', Warning)
+                    warnings.warn('trestle remove of a list removes the entire list', Warning)
+                elif type(deleting_element) is dict:
+                    pass
+                    warnings.warn('trestle remove does not support removing dict elements', Warning)
+                    warnings.warn('trestle remove of a dict element removes the entire dict element', Warning)
```
Warnings give a strange user experience. Use logger.warn(). Not that trestle has a logging handler which is centralised using trestle.utils.log.get_logger()

jeffdmgit replied:
```
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.backmatter
Remove failed: Bad element path: catalog.backmatter
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata
Remove failed: Last stream position is not available to rollback to
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata.*
Remove failed: trestle remove does not support Wildcard element path.
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$
```

The handling of a validation failure from a plan simulation is not ideal, but the actual error, i.e., removing a required element in this case, is not properly passed back by pydantic, which normally throws:
```
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/tr
estle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata
Traceback (most recent call last):
  File "/Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/core/models/elements.py", line 359, in set_at
    setattr(preceding_elm, sub_element_name, model_obj)
  File "pydantic/main.py", line 394, in pydantic.main.BaseModel.__setattr__
pydantic.error_wrappers.ValidationError: 1 validation error for Catalog
metadata
  none is not an allowed value (type=type_error.none.not_allowed)
```

So for now it's not a useful error message from simulate(). I do note that it comes from simulate() in the log, but not in the final error returned.

And yes, I am now using a logger instead of warnings.

butler54 commented on trestle/core/commands/remove.py:
```diff
@@ -15,10 +15,113 @@
+        The method first finds the parent model from the file and loads the file into the model.
+        Then the method executes 'remove' for each of the element paths specified.
+        """
+        args = args.__dict__
```
You don't need to use args.__dict__. You can just use args.args_name.

jeffdmgit replied:
I'd have to use _argparse_, which is not used in **trestle add** either. Perhaps we can consider bringing in _argparse_ in a separate issue.

butler54 replied:
The add arguments are argparse arguments. Check out trestle create for an example.

butler54 commented on trestle/core/commands/remove.py:
```diff
@@ -15,10 +15,113 @@
+            required=True
+        )
+
+    def _run(self, args):
```
If you could add some type definitions to the function definitions it would be fantastic.
e.g. 
```python
def _run(self, args: argparse.Namespace) -> int: 

```
A further note here - this function can pass back a return code. To date we have been using 0 for pass and 1 for any error (e.g. errors are untyped).

butler54 replied:
In reality this happens already to some extent by exception - but testing this function specifically allows you to check that exceptions are not been thrown (as opposed to sys.exit()).

jeffdmgit replied:
Done, added.

butler54 reviewed (changes requested):
So I've found a few errors which are unhandled. 
All the references below assume that `catalog.json` is nist 800-53 rev 5.

1) Remove fails with a bad element path by throwing an exception rather than an error message:[
```
trestle remove -f catalog.json -e 'catalog.backmatter'
Traceback (most recent call last):
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 120, in remove
    raise err.TrestleError(f'Bad element path. {str(element_path)}')
trestle.core.err.TrestleError: Bad element path. catalog.backmatter

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/chris/opt/anaconda3/bin/trestle", line 33, in <module>
    sys.exit(load_entry_point('compliance-trestle', 'console_scripts', 'trestle')())
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/cli.py", line 52, in run
    exit(Trestle().run())
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 147, in run
    return parsed_args.func(parsed_args, extra_args=extra_args)
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 174, in _validate_and_run
    return self._run(parsed_args)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 74, in _run
    remove_action, parent_element = self.remove(element_path, parent_model, parent_element)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 123, in remove
    raise err.TrestleError(f'Bad element path. {str(e)}')
trestle.core.err.TrestleError: Bad element path. Bad element path. catalog.backmatter

2) trestle remove fails badly when trying to remove a required element.
```
trestle remove -f catalog.json -e 'catalog.backmatter'
Traceback (most recent call last):
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 120, in remove
    raise err.TrestleError(f'Bad element path. {str(element_path)}')
trestle.core.err.TrestleError: Bad element path. catalog.backmatter

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/chris/opt/anaconda3/bin/trestle", line 33, in <module>
    sys.exit(load_entry_point('compliance-trestle', 'console_scripts', 'trestle')())
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/cli.py", line 52, in run
    exit(Trestle().run())
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 147, in run
    return parsed_args.func(parsed_args, extra_args=extra_args)
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 174, in _validate_and_run
    return self._run(parsed_args)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 74, in _run
    remove_action, parent_element = self.remove(element_path, parent_model, parent_element)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 123, in remove
    raise err.TrestleError(f'Bad element path. {str(e)}')
trestle.core.err.TrestleError: Bad element path. Bad element path. catalog.backmatter
(base) chris@jettopper NIST_2 % trestle remove -f catalog.json -e 'catalog.back-matter'
(base) chris@jettopper NIST_2 % trestle remove -f catalog.json -e 'catalog.metadata'
Traceback (most recent call last):
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/elements.py", line 359, in set_at
    setattr(preceding_elm, sub_element_name, model_obj)
  File "pydantic/main.py", line 394, in pydantic.main.BaseModel.__setattr__
pydantic.error_wrappers.ValidationError: 1 validation error for Catalog
metadata
  none is not an allowed value (type=type_error.none.not_allowed)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/plans.py", line 69, in simulate
    raise ex
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/plans.py", line 67, in simulate
    self.execute()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/plans.py", line 76, in execute
    action.execute()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/actions.py", line 425, in execute
    self._src_element.set_at(self._sub_element_path, None)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/elements.py", line 364, in set_at
    but found "{model_obj.__class__}"'
trestle.core.err.TrestleError: Validation error: metadata is expected to be "<class 'trestle.oscal.catalog.Metadata'>",                     but found "<class 'NoneType'>"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/chris/opt/anaconda3/bin/trestle", line 33, in <module>
    sys.exit(load_entry_point('compliance-trestle', 'console_scripts', 'trestle')())
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/cli.py", line 52, in run
    exit(Trestle().run())
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 147, in run
    return parsed_args.func(parsed_args, extra_args=extra_args)
  File "/Users/chris/opt/anaconda3/lib/python3.7/site-packages/ilcli/command.py", line 174, in _validate_and_run
    return self._run(parsed_args)
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/commands/remove.py", line 84, in _run
    add_plan.simulate()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/plans.py", line 71, in simulate
    self.rollback()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/plans.py", line 84, in rollback
    action.rollback()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/actions.py", line 210, in rollback
    super().rollback()
  File "/Users/chris/go/src/github.com/IBM/compliance-trestle/trestle/core/models/actions.py", line 155, in rollback
    raise TrestleError('Last stream position is not available to rollback to')
trestle.core.err.TrestleError: Last stream position is not available to rollback to
```


```

jeffdmgit commented:
Not quite 100% after the unhandled exceptions were resolved in remove.py.

With the fixes, and added tests, I’m up 95% coverage now, and fixed something that was previously there. The last 5% will probably require a bit more than the prior patterns because simulate() actually catches the problem before execute() does, for a bad remove plan. I think I’ll have to figure out how to feed a bad plan to it.

jeffdmgit commented:
This is now done with 100% coverage.

butler54 reviewed (approved):
LGTM - please see comment on future improvements.
